libflux: fix composite future implementation

grondo opened the pull request:
Unfortunately, the composite future implementation was committed with only sanity unit tests, and as was determined by @garlick in #1750, the current code was broken for use with RPCs (and probably any other real-world usage).

The problem is that current executing context of the future on which the user may call `flux_future_then(3)` or `flux_future_wait_for(3)` was not properly propagated to all futures that may be involved in the composite. E.g. if a user composes a future from several back-to-back RPCs, then calls a blocking `flux_future_get(3)` on the composite, the `get()` will be sleeping on the "now" context reactor and associated "cloned" `flux_t` handle, but the predecessor futures will by default use the "main" `flux_t` handle, and thus those responses will be parked until `flux_future_get(3)` returns and the chain deadlocks.

The "fix" is to be sure to propagate the current context (reactor and flux_t handle, if any) to other composites children or prev/next futures at key points (creation, init callback, and flux_future_continue).

Also, it was apparent that the reference handling for `flux_future_continue(3)` wasn't clear from the docs, so this was clarified in the manpage and the `future.h` header.

There's some incremental development on `t/rpc/chained.c` here to make it easy to review. If the changes are ok, then we can squash that before merging.

Finally, it would be nice to flesh out the rest of the tests here (still missing tests for `flux_future_or_then(3)` and `flux_future_wait_any/all(3)`. I can do that in a future PR if we just want to get these fixes in  now.

garlick commented:
Nice!  Thanks so much for fixing this.  I was really perplexed when I tried earlier.

Feel free to squash those test additions into the original commit.

Commit message for  97f5d5b contains a spelling error: becaause

I would vote we get this in ASAP and add more tests later since this is damned useful.

grondo commented:
Ok, squashed, fixed typo, and rebased on current master.

I also had time to at least add sanity check in "now" context for `flux_future_or_then(3)` to the new unit test.
